exporters/otlp/retry.go

```diff
@@ -0,0 +1,63 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlp
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+// RetrySettings defines configuration for retrying batches in case of export failure.
+// The current supported strategy is exponential backoff.
+type RetrySettings struct {
+	// Enabled indicates whether to not retry sending batches in case of export failure.
+	Enabled bool
+	// InitialInterval the time to wait after the first failure before retrying.
+	InitialInterval time.Duration
+	// MaxInterval is the upper bound on backoff interval. Once this value is reached the delay between
+	// consecutive retries will always be `MaxInterval`.
+	MaxInterval time.Duration
+	// MaxElapsedTime is the maximum amount of time (including retries) spent trying to send a request/batch.
+	// Once this value is reached, the data is discarded.
+	MaxElapsedTime time.Duration
+}
+
+// DefaultRetrySettings returns the default settings for RetrySettings.
+func DefaultRetrySettings() RetrySettings {
+	return RetrySettings{
+		Enabled:         true,
+		InitialInterval: 5 * time.Second,
+		MaxInterval:     30 * time.Second,
+		MaxElapsedTime:  time.Minute,
+	}
+}
+
+func NewExponentialConfig(rs RetrySettings) *backoff.ExponentialBackOff {
```

Comment:
This shouldn't be exported, there is no reason a user should be calling this.

```suggestion
func newExponentialConfig(rs RetrySettings) *backoff.ExponentialBackOff {
```

Reply:
This doesn't return a config, it returns an object that has state and other properties.
```suggestion
func newExponentialBackoff(rs RetrySettings) *backoff.ExponentialBackOff {
```

Reply:
I've moved this to the package it is used and unexported it. Initially, I've added it to the `otlp` so both the `otlpgrpc` and `otlphttp` packages could use it, but I'm scoping it to grpc only at the moment. If we plan to use this code in the `otlphttp` it would easy to refactor it.